hub/core/chunk_engine.py

```diff
@@ -0,0 +1,315 @@
+from hub.core.compression import decompress_array
+from math import ceil
+from typing import Optional, Sequence, Union, Tuple
+from hub.util.exceptions import (
+    CorruptedMetaError,
+    DynamicTensorNumpyError,
+    TensorDoesNotExistError,
+)
+from hub.core.meta.tensor_meta import TensorMeta
+from hub.core.index.index import Index
+from hub.util.keys import (
+    get_chunk_key,
+    get_chunk_id_encoder_key,
+    get_tensor_meta_key,
+)
+from hub.core.sample import Sample
+from hub.constants import DEFAULT_MAX_CHUNK_SIZE, UNCOMPRESSED
+
+import numpy as np
+
+from hub.core.storage.lru_cache import LRUCache
+
+from hub.core.chunk import Chunk
+
+from hub.core.meta.encode.chunk_id import ChunkIdEncoder
+
+
+SampleValue = Union[np.ndarray, int, float, bool, Sample]
+
+
+def is_uniform_sequence(samples):
+    if len(set(map(type, samples))) != 1:
+        # Cannot vectorize sequence with inconsistent types
+        return False
+    elif any(isinstance(s, np.ndarray) for s in samples):
+        # Numpy arrays will only be vectorized if they have the same shape
+        return len(set(s.shape for s in samples)) == 1
+    elif any(isinstance(s, Sample) for s in samples):
+        # Sample objects will not be vectorized
+        return False
+    else:
+        # Scalar samples can be vectorized
+        return True
+
+
+class ChunkEngine:
+    def __init__(
+        self, key: str, cache: LRUCache, max_chunk_size: int = DEFAULT_MAX_CHUNK_SIZE
+    ):
+        """Handles creating chunks and filling them with incoming samples.
+
+        Args:
+            key (str): Tensor key.
+            cache (LRUCache): Cache for which chunks and the meta information are stored.
+            max_chunk_size (int): Chunks generated by this instance will never exceed this size. Defaults to DEFAULT_MAX_CHUNK_SIZE.
+
+        Raises:
+            ValueError: If invalid max chunk size.
+        """
+
+        # TODO: elaborate further in docstring (including our encoders)
+        # TODO: add examples
+
+        self.key = key
+        self.cache = cache
+
+        if max_chunk_size <= 2:
+            raise ValueError("Max chunk size should be > 2 bytes.")
+
+        self.max_chunk_size = max_chunk_size
+        self.min_chunk_size_target = self.max_chunk_size // 2
```

Comment:
@AbhinavTuli looks like this is what happened to `CHUNK_MIN_SIZE`

Reply:
yeah i think this is better, requires only 1 parameter, and half is a really good heuristic here